File: 2019/03/go/segment.go
package main

import (
	"fmt"
	"strconv"
)

type Segment struct {
	Points [2]Point
}

func NewSegment(start Point, str string) (*Segment, error) {
	if str == "" {
		return nil, fmt.Errorf("Empty source string")
	}

	magnitude, err := strconv.ParseInt(str[1:], 10, 32)
	if err != nil {
		return nil, err
	}

	segment := Segment{}
	segment.Points[0] = start

	switch str[0] {
	case 'U':
		segment.Points[1] = Point{start.X, start.Y + magnitude}
	case 'D':
		segment.Points[1] = Point{start.X, start.Y - magnitude}
	case 'L':
		segment.Points[1] = Point{start.X - magnitude, start.Y}
	case 'R':
		segment.Points[1] = Point{start.X + magnitude, start.Y}
	default:
		return nil, fmt.Errorf("Invalid character in source string start: %v", str[0])
	}

	return &segment, nil
}

func (s Segment) IsVertical() bool {
	if s.Points[0].X == s.Points[1].X {
		return true
	} else {
		return false
	}
}

func (s Segment) IsHorizontal() bool {
	if s.Points[0].Y == s.Points[1].Y {
		return true
	} else {
		return false
	}
}

File: 2019/03/zig/segment.zig
//------------------------------------------------------------------------------
// GrooveStomp's Advent of Code
// Copyright (c) 2019 Aaron Oman (GrooveStomp)
//
// File: segment.zig
// Created: 2019-12-10
// Updated: 2019-12-10
// Author: Aaron Oman (GrooveStomp)
//
// Notice: CC BY 4.0 License
//
// This program comes with ABSOLUTELY NO WARRANTY.
// This is free software, and you are welcome to redistribute it under certain
// conditions; See https://creativecommons.org/licenses/by/4.0/ for details.
//------------------------------------------------------------------------------
const std = @import("std");
const p = @import("point.zig");
const alloc = std.heap.direct_allocator;
const warn = std.debug.warn;
const math = std.math;

pub const Segment = struct {
    points: [2]p.Point,

    pub fn isVertical(s: Segment) bool {
        return s.points[0].x == s.points[1].x;
    }

    pub fn isHorizontal(s: Segment) bool {
        return s.points[0].y == s.points[1].y;
    }
};

pub fn NewSegment(point: p.Point, str: []u8) !*Segment {
    if (std.mem.eql(u8, str, "")) {
        return error.Unexpected;
    }

    var magnitude = @intCast(isize, try std.fmt.parseInt(usize, str[1..], 10));
    var segment = try alloc.create(Segment);
    segment.points[0] = point;

    switch (str[0]) {
        'U' => segment.points[1] = p.Point{ .x = point.x, .y = point.y + magnitude },
        'D' => segment.points[1] = p.Point{ .x = point.x, .y = point.y - magnitude },
        'L' => segment.points[1] = p.Point{ .x = point.x - magnitude, .y = point.y },
        'R' => segment.points[1] = p.Point{ .x = point.x + magnitude, .y = point.y },
        else => {
            warn("Unknown direction\n");
            return error.Unexpected;
        }
    }

    return segment;
}
